### ISSUE TITLE
Consider possibility of nightly/snapshot updates in future npm release

### ISSUE THREAD

**Issue by Beak-man** [NASAWorldWind/WebWorldWind#141]
Closes #136

**Comment by pdavidc** [comment]
Recommend limiting this investigation to 0.5 days.

**Comment by pdavidc** [comment]
I came across a post describing how the Angular project has been publishing nightly/snapshot builds via NPM.
https://blog.lacolaco.net/post/how-to-use-angular-2-nightly-builds/

**Comment by pdavidc** [comment]
We left the choice of nightlies vs snapshots open on this issue to accommodate any existing JavaScript conventions we've yet to discover. The CyanogenMod definition of snapshot is not a common one, and certainly not what we intended.

What would help WorldWind users the most is to have a snapshot deployed on every push to develop. That capability enables immediate deployment of prototype changes and bug fixes. Is there a good enough convention for snapshots that's in use by JavaScript developers that we could explain and teach to WorldWind users?

**Comment by zglueck** [comment]
@pdavidc I used the GitHub dependency specification you demonstrated in your link and is formally documented in the [npm docs](https://docs.npmjs.com/files/package.json#github-urls) to test using the WebWorldWind GitHub repo as a "SNAPSHOT" source for the latest features.

I leveraged @Beak-man react test app and removed the `package.json` dependency on the tar ball and instead used:
```
npm install --save "NASAWorldWind/WebWorldWind#develop"
```

Unfortunately, this doesn't work. I think because the file specified by the `main` property, `worldwind.min.js`, doesn't exist in the repository. I forked the WebWorldWind repo and added the minified library and then it worked.

@Beak-man am I missing anything here? I browsed some other npm projects (angular, express) and all of them seem to have an `index.js` in the repository.

**Comment by pdavidc** [comment]
The WebWorldWind packages entry point, worldwind.min.js, is expected to be built prior to publication. I'm sure you're familiar with the WebWW build system. Is this a case of something simple overlooked, or is there a problem executing the build in the environment you're describing?

**Comment by zglueck** [comment]
I think its fine when published, but to utilize as a "SNAPSHOT" like dependency using a GitHub repository, it looks like npm copies the GitHub repository contents which doesn't include the built library. This is the error received in React test app:

```
./src/WorldWindComponent.js
Module not found: Can't resolve '@nasaworldwind/worldwind' in '/private/tmp/worldwind-react/src'
```

Take a look at the [express](https://github.com/expressjs/express) module. The npm documentation uses express as an example and it includes the index.js file in the repo. I think it is just an architecture difference from the original development of Web WorldWind.

This could be a point where the `post-install` scripts may be worthwhile. I haven't tested yet, but if they build the library post install using the GitHub reference, then the module would be available.

**Comment by pdavidc** [comment]
I'm unfamiliar with the method to provide snapshots that you've found. You're the first to really investigate what's possible here, so it would be helpful if you'd outline the method and your thoughts from a high level.

**Comment by zglueck** [comment]
Our goal is to provide users a method to get the latest code from the develop branch, similar to the SNAPSHOT artifact of java. As @Beak-man noted, modules published to the npm repository utilize semver and immutable modules. Just updating the code without updating the version is prohibited.

Fortunately, npm allows for specifying [GitHub repositories as dependencies](https://docs.npmjs.com/files/package.json#github-urls). When specified like this, the strict versioning requirement is dropped as the `npm install` command will now clone the contents of the GitHub repository into the node_modules directory alongside the more conventional npm repository modules. This is the approach outlined in https://blog.lacolaco.net/post/how-to-use-angular-2-nightly-builds/

Our project setup requires a build process to produce our exported module (worldwind.min.js). In a publishing workflow, this isn't an issue, as we will build before publishing, and then the module will be published. When the repository is used as a GitHub dependency, the requirement to build the library isn't currently part of the cloning/download process. [I believe adding a `postinstall` may allow the module to be built after local install](https://docs.npmjs.com/misc/scripts). The `postinstall` script would build the module on a local install.

So to summarize, there is a path forward to providing "SNAPSHOT" development modules to the general public in a common npm method. Our repository/project setup currently doesn't facilitate this use as our exported module depends on a build process. npm scripting options may be a solution by building the exported module after local install.

**Comment by pdavidc** [comment]
Thanks!

I'll look forward to your investigation into the npm scripting option.

**Comment by zglueck** [comment]
I have not found any documentation on how npm is aware of updates to the GitHub dependency. Does npm automatically check the GitHub repository for changes and download automatically (similar to the behavior of a java SNAPSHOT)? Or is there a manual update requirement?

I'll add this to the investigation.

**Comment by zglueck** [comment]
@pdavidc @AkeluX and @Beak-man 

After working through this, I'm starting to question whether `worldwind.js` should be specified as the `main` entry point instead of `worldwind.min.js`.

If a project is using npm with WorldWind as a dependency, isn't a bundler a likely part of the project already? The `worldwind.js` file would be minified alongside the rest of the dependencies and application by the bundler. Isn't bundling and minification really a task of the project using WorldWind?

This discussion extends further, should the main entry point be `WorldWind.js`? I know there is a bit of module definition craziness switching between AMD and CommonJS so I'm not sure this is possible.

My main point, should our `main` property point at a module which requires a build process? Particularly because an app using the project will likely have a bundling function of its own.

To be clear, I'm not advocating for the removal of the infrastructure to create the worldwind.js and worldwind.min.js files as they have use outside the npm ecosystem, but merely trying to separate concerns.

**Comment by pdavidc** [comment]
These are good points, but they complicate this issue and expand its scope unnecessarily. Stick to the objective of this issue.

**Comment by pdavidc** [comment]
@zglueck, I've been thinking about the questions you raised and come to the conclusion that the answers may be critical to how we do snapshots.

That said, we didn't give this task a clear objective from the beginning, is it's no surprise that it's unclear what to do here. Let's treat this as a spike with the following objectives:

- Determine how npm snapshots could work. ( you've already done this )
- Identify what would need to change to support the snapshotting method.

**Comment by Beak-man** [comment]
@zglueck:
I never tried using `WorldWind.js` as the main entry point, assuming that it wouldn't work precisely for the same reasons you have stated. That being said, I assumed that because setting either of the built artifacts as main entry points before adding a CommonJS interface to the whole library in #151 was one of the first things I tried, back when the library was able to only interface as an AMD module (or an object attached to `window`). Of course, it didn't work.

Regarding the usage of the unminified or minified version of the library, that's a good question. I chose to go with the minified one in case the developer is testing the performance of WebWW's npm package in isolation before bundling their project.

And yeah, most npm packages have `index.js` as entry points, which alongside with jszip's and proj4's wonky Browserify-made UMD interfacing, got me stuck on scoping the package for a while, which suggests that npm makes some assumptions when the entry point is `index.js`. Consider that it's relatively rare to see UI libraries that were repurposed as npm packages (without being rewritten, in CommonJS modules with Browserify, for instance), and pretty much everyone begins developing a package with an `index.js` file now. 

But it works when `worldwind.min.js` is added to the repo. @pdavidc, do you think it would be worthwhile to include the built artifacts to the repository? (in a '/dist' folder, for instance). I see that many projects do that (e.g. [jszip itself](https://github.com/Stuk/jszip/tree/master/dist)).

**Comment by pdavidc** [comment]
As I mentioned above, these are important points, but they're out of scope for this task.

**Comment by zglueck** [comment]
## WebWorldWind Development Builds
npm's dependency resolution flexibility provides a few different options for development builds. I encourage a review of [npm package.json documentation](https://docs.npmjs.com/files/package.json#dependencies) for a broad overview of the dependency specifications.

Many options are available for providing development builds, some of the best options are detailed below.

### Tarball URL Hosting
The npm package documentation states a URL to a tarball (the tarball as created by the `npm pack` command) may be used as a dependency. Users of the development build would specify the url in their project's package.json file. The effort on our end would require hosting of the tarball and publishing the updated tarball on development push events (similar to WWA's behavior). The building and publishing tasks are perfectly suitable for Travis. A URL and hosting service would need to be secured.

Changes Required:
```
- Secure hosting for tarball (external)
- Integrate build and publish to hosting in Travis (.travis.yml)
```

### GitHub Hosting
Similar to the URL hosting, users may also specify a GitHub repository and branch in the dependencies. This approach currently does not work because the `main` property specifies a file which requires a build process. npm actually clones the repository, but as currently configured, conducts no build process.

There are a few ways around this limitation:

1. As @Beak-man noted, include the built artifacts in the repository, thus requiring no build process client side. While this is a simple and quick solution, I think the hang up is in how the built artifact is updated in the repository. There are many options which warrant a separate and deep discussion if this path is desired.

Changes Required:
```
- Configure how artifact is added to repository (???, .gitignore)
```

2. Utilize the [postinstall-build](https://www.npmjs.com/package/postinstall-build) package which adds a `postinstall` script which npm will execute after installing the package. The `postinstall-build` package is convenient because it doesn't require all of your `devDependencies` to become `dependencies` in order to build on the client machine. It can be configured to run specific scripts and only run them when the package is downloaded as a dependency. Testing revealed some sensitivity to npm versions and the maintainer of the package officially acknowledges some versions of npm are not compatible. The `postinstall-build` package would need to become a `dependency` of the project.

Changes Required:
```
- $ npm install --save postinstall-build (command line)
- "postinstall": "postinstall-build worldwind.js --script=bundle --only-as-dependency"
 (package.json)
```

3. Implement an `index.js` shim which would load `WorldWind.js`. I attempted a few different approaches without success. Someone with a better grasp of AMD and CommonJS interactions would likely have more success. As noted by @pdavidc, this is larger topic outside the scope of this task but wanted to document, if successful, it would allow GitHub development artifacts.

Changes Required:
```
- Add index.js with shim
- Likely need to add requirejs as a dependency
```

## Update Frequency

I wasn’t able to track down documentation indicating how often (if ever) npm checked packages. Testing indicated running `npm update` did update the node_modules version of the library with the latest GitHub contents.

## Conclusion

There a number of options for providing development builds to the community but no immediate turn-key solutions.
